packages/dev/sharedUiComponents/src/historyStack.ts

```diff
@@ -133,26 +130,20 @@ export class HistoryStack implements IDisposable {
             return;
         }
 
-        const data = this._dataProvider();
-        const dataString = JSON.stringify(data);
+        const data = this._copy(this._dataProvider());
 
-        if (this._history.length > 0) {
-            const previousState = this._rebuildState();
-            if (JSON.stringify(previousState) === dataString) {
+        if (this._activeData) {
+            const diff = this._generateJSONDiff(data, this._activeData);
+            if (!diff) {
                 return;
             }
-            const diff = this._generateJSONDiff(previousState, data);
-            if (diff) {
-                this._history.push(JSON.stringify(diff));
-            }
-        } else {
-            this._history.push(dataString);
+            this._historyStack.push(JSON.stringify(diff));

```

Comment:
Why not storing directly `diff`? It would avoid the `JSON.stringify` call when adding to the history stack, and `JSON.Parse` when retrieving the data.